### ISSUE TITLE
iOS Auto-Fill for password fields

### ISSUE THREAD

**Issue by hpoul** [authpass/authpass#31]
https://developer.apple.com/documentation/authenticationservices

**Comment by hpoul** [comment]
This seems to be a bit more complicated than i anticipated. the auto fill can only run in an app extension (autofill provider), which has no way to communicate with the main app (except through keychain).. and there is no easy way to launch a flutter view inside an app extension.. 

so the only feasible way i could think of would be to sync all passwords from the kdbx into a shared keychain.. but i'm not sure if this is a good idea security and feature wise.

i set it to a low priority for now.. but just give it a 👍️ if anyone is interested.. if there is some demand maybe it's worth investigating further.

**Comment by hpoul** [comment]
I think I've got an idea for an intermediary solution/workaround. How about slowing to launch a website directly from AuthPass and temporarily transfer the password for that entry into the autofill app extension - then the user could autofill that one entry directly into the website..
Anyone thinks that this would be a worthwhile workaround?

**Comment by OGmetamonkey** [comment]
I think your proposal to sync all passwords to an autofill provider would be the most appropriate  implementation. I think this is possible using the [credential identity store](https://developer.apple.com/documentation/authenticationservices/ascredentialidentitystore) API. Good walkthrough here: [walkthrough](https://developer.apple.com/videos/play/wwdc2018/721/)

**Comment by abuisman** [comment]
I’d love for this to work on iOS. 

What is currently stopping this from working? Are there any developments?

**Comment by hpoul** [comment]
@abuisman that would be great.. I haven't done anything on that topic for a year.. https://github.com/authpass/authpass/commit/5cb739007189a1419a8a1b54485484c854dd9023 was my last try.. I don't think there is much to salvage from that :-(

The biggest blocker is that it's not trivial to get the flutter engine running in an app extension.. I think the safest bet would be to not try that.. 😅️ I'm not completely sure about the security implications.. but the best way to start would probably be to simply sync all accounts into a iOS keychain .. and have the autofill app extension (ie AutoFill Credential Provider Extensions) look up the accounts in the keychain.. and don't even try to work with theunderlyingg kdbx file.

this is also basically how apple presents it in their demo.. https://developer.apple.com/videos/play/wwdc2018/721/

<img width="321" alt="Screen Shot 2021-08-10 at 09 27 03" src="https://user-images.githubusercontent.com/313066/128825999-370c216c-58eb-4278-ba46-e26f50db4294.png">

I guess firing up a whole flutter engine in an app extension is not a good idea anyway..

**Comment by hpoul** [comment]
For reference - here is the flutter issue related to having a flutter engine in an app extension: https://github.com/flutter/flutter/issues/16092

**Comment by abuisman** [comment]
Haha I didn’t mean “I’d love to work on this” I meant it would be great if this feature was functional on iOS. I don’t think I’d be the right person for this as I have neither flutter or iOS experience. Let alone I have no Apple developer account 😇

**Comment by abuisman** [comment]
Would the bitwarden source code be useful? They have this functionality working quite well.

**Comment by hpoul** [comment]
> Would the bitwarden source code be useful

no

**Comment by softrider70** [comment]
The other password safes under ios like Authenticator or msecure are registered in ‚automatsch ausfüllen‘, so maybe you need the methodes behind this, to implement the autofill.
Would be great if this works.

**Comment by abuisman** [comment]
> > Would the bitwarden source code be useful
> 
> 
> 
> no

**Comment by oliverbytes** [comment]
> This seems to be a bit more complicated than i anticipated. the auto fill can only run in an app extension (autofill provider), which has no way to communicate with the main app (except through keychain).. and there is no easy way to launch a flutter view inside an app extension..
> 
> so the only feasible way i could think of would be to sync all passwords from the kdbx into a shared keychain.. but i'm not sure if this is a good idea security and feature wise.
> 
> i set it to a low priority for now.. but just give it a 👍️ if anyone is interested.. if there is some demand maybe it's worth investigating further.

When do you plan to start working on syncing with keychain instead? Seems the only way right now! Kudos for your great work!

**Comment by FnAndroid** [comment]
> I think I've got an idea for an intermediary solution/workaround. How about slowing to launch a website directly from AuthPass and temporarily transfer the password for that entry into the autofill app extension - then the user could autofill that one entry directly into the website.. Anyone thinks that this would be a worthwhile workaround?


Can I choose to write an autofill app extension using Objective-C or webSite? AuthPass is the best open source password management tool, but at the moment it lacks the most important autofill feature.

**Comment by oliverbytes** [comment]
> > I think I've got an idea for an intermediary solution/workaround. How about slowing to launch a website directly from AuthPass and temporarily transfer the password for that entry into the autofill app extension - then the user could autofill that one entry directly into the website.. Anyone thinks that this would be a worthwhile workaround?
> 
> 
> Can I choose to write an autofill app extension using Objective-C or webSite? AuthPass is the best open source password management tool, but at the moment it lacks the most important autofill feature.

A separate native iOS AuthPass extension would work I believe

**Comment by kmkmahesh** [comment]
I use [Keepass Touch](https://apps.apple.com/us/app/keepass-touch/id966759076) which autofills the passwords

**Comment by oliverbytes** [comment]
Someone was able to run flutter as iOS extension: https://github.com/flutter/flutter/issues/16092?notification_referrer_id=MDE4Ok5vdGlmaWNhdGlvblRocmVhZDMxOTExNDEyNTozOTYzNjY%3D#issuecomment-1288192913

**Comment by hpoul** [comment]
@oliverbytes yep, I've seen that.. but it requires a custom patch to run.. and even if that using gets merged into flutter it's not enough to get accepted into the apple app store as far as i can tell.. but we'll see, maybe one day.... 😅
But if someone wants to develop a native ios app extension, I'd be happy to review it help with a PR

**Comment by nexbotai** [comment]
Will it be possible now that iOS App Extensions are officially supported? https://docs.flutter.dev/platform-integration/ios/app-extensions

**Comment by hpoul** [comment]
@nexbotai not really.. the problem is still how the app extension could access the passwords.. the app does not run at the same time as the extension.. imho the only/best way would be to have the app sync all passwords  into a shared keychain.. 
Or maybe try to use flutter in the app extension itself so it can read the kdbx itself.. but that would require a custom built if the flutter engine.. which sounds pretty annoying.. 
But if anyone wants to give it a try... Feel free

**Comment by hpoul** [comment]
fwiw.. i think in the meantime it is actually possible to run flutter inside an app extension.. (although development is still pretty buggy).. but it's a first step.. will take a while until it gets usable, and not sure how much time i've got to play around with it

https://github.com/authpass/authpass/assets/313066/ca07dc3f-5dd7-4205-9176-3a6d7b888594

**Comment by Dige006** [comment]
Hello Herbert,
that sounds very promising.

As far as I have been able to see so far, the missing autofill function is the only (important) feature that prevents me from using it.

I would be very happy if it continues here and would like to thank you very much for your work.

**Comment by qmzmdqs** [comment]
Still looking forward to this important feature,
perhaps a native ios implementation would be the most convenient and reliable!

**Comment by Nvdtn19** [comment]
This problem is still not has been solved yet? This missing important feature is the largest thing that prevent me from using this password manager so far, despite this password manager has many useful features. Or if this feature is still in beta, please at least make a chance for beta tester to test this feature and give the feedback. If anyone somehow managed to implement this feature, I'd appreciate if they can share their method.

**Comment by yuvalnbreachads** [comment]
Any news?

**Comment by hpoul** [comment]
There is now actually a test implementation in #415 . If anyone feels adventurous give it a spin.. It's released to TestFlight https://testflight.apple.com/join/3CC54i3V

**Comment by Nvdtn19** [comment]
> There is now actually a test implementation in [#415](https://github.com/authpass/authpass/pull/415) . If anyone feels adventurous give it a spin.. It's released to TestFlight https://testflight.apple.com/join/3CC54i3V

Is there still a valid TestFlight invite? I can't visit this

**Comment by hpoul** [comment]
@Nvdtn19 ups, sorry you are right, forgot to publish it to the "Public Testers" group. It now needs to go through Beta-Review, should take a couple of hours... I'll let you know once it's through.

**Comment by hpoul** [comment]
> > There is now actually a test implementation in [#415](https://github.com/authpass/authpass/pull/415) . If anyone feels adventurous give it a spin.. It's released to TestFlight https://testflight.apple.com/join/3CC54i3V
> 
> Is there still a valid TestFlight invite? I can't visit this
> 

@Nvdtn19 beta app review is now approved, so joining TestFlight is possible again.
